sql/core/src/test/scala/org/apache/spark/sql/sources/v2/DataSourceV2SQLSuite.scala

```diff
@@ -170,6 +173,85 @@ class DataSourceV2SQLSuite extends QueryTest with SharedSQLContext with BeforeAn
     checkAnswer(spark.internalCreateDataFrame(rdd, table.schema), spark.table("source"))
   }
 
+  test("ReplaceTableAsSelect: basic v2 implementation using atomic catalog.") {
+    spark.sql("CREATE TABLE testcatatomic.table_name USING foo AS SELECT id, data FROM source")
+    val testCatalog = spark.catalog("testcatatomic").asTableCatalog
+    val originalTable = testCatalog.loadTable(Identifier.of(Array(), "table_name"))
+    spark.sql("REPLACE TABLE testcatatomic.table_name USING foo AS SELECT id FROM source")
+    val replacedTable = testCatalog.loadTable(Identifier.of(Array(), "table_name"))
+    assert(replacedTable != originalTable, "Table should have been replaced.")
+    assert(replacedTable.name == "testcatatomic.table_name")
+    assert(replacedTable.partitioning.isEmpty)
+    assert(replacedTable.properties == Map("provider" -> "foo").asJava)
+    assert(replacedTable.schema == new StructType()
+      .add("id", LongType, nullable = false))
+    val rdd = spark.sparkContext.parallelize(replacedTable.asInstanceOf[InMemoryTable].rows)
+    checkAnswer(
+        spark.internalCreateDataFrame(rdd, replacedTable.schema),
+        spark.table("source").select("id"))
+  }
+
+  test("ReplaceTableAsSelect: Non-atomic catalog creates the empty table, but leaves the" +
+    " table empty if the write fails.") {
```

Comment:
Why isn't the table dropped in this case? I would expect this to have the behavior of non-atomic CTAS after the initial delete.

Reply:
I think the behavior here is ambiguous. Suppose then another user went and started writing to the table concurrently - should this job drop the table that the other job is writing to?

Reply:
I think the intent of RTAS is to run `DROP TABLE` and `CREATE TABLE ... AS SELECT ...`. CTAS doesn't worry about concurrent writes because the table doesn't "exist" until the write completes. That's why we delete after a CTAS if the write fails, even though it also has a non-atomic case where the table exists and could technically be written to concurrently.

Reply:
Hm so I took a closer look at this. It turns out that using `Utils.tryWithSafeFinallyAndFailureCallbacks` is risky in all of these code paths as it is currently implemented.

That method, when it tries to run the catch block, will first try to set the failure reason on the task context via TaskContext.get().markTaskFailed`. But since we're running this try...finally block on the _driver_, there is no such task context to get via `TaskContext.get`.

What happens in this case then is that this test passes when it should fail, because indeed, the table should be dropped. But the catch block that drops the table never gets run, because `TaskContext.get().markTaskFailed` NPEs before the catch block can be run.

I think there's a few ways forward:
1) Don't use the Utils method to do try-catch-finally
2) Patch the Utils method to check for null on the current task context before trying to mark the task failure reason on it.

I'm going with 2) for now, but 1) is very reasonable as well.

Either way, yeah the table should end up being dropped at the end, so this test also has to be patched.